internal/daemon/api_files.go

```diff
@@ -0,0 +1,627 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package daemon
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"net/textproto"
+	"os"
+	"os/user"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+)
+
+func v1GetFiles(_ *Command, req *http.Request, _ *userState) Response {
+	query := req.URL.Query()
+	action := query.Get("action")
+	switch action {
+	case "read":
+		paths := query["path"]
+		if len(paths) == 0 {
+			return statusBadRequest("must specify one or more paths")
+		}
+		return readFilesResponse{paths: paths}
+	case "list":
+		pattern := query.Get("pattern")
+		if pattern == "" {
+			return statusBadRequest("must specify pattern")
+		}
+		directory := query.Get("directory")
+		if directory != "true" && directory != "false" && directory != "" {
+			return statusBadRequest(`directory parameter must be "true" or "false"`)
```

Comment:
It's equivalent to `ls -d`, which shows the directory itself rather than what's inside the directory. Compare:

```
$ ls -al -d /etc
drwxr-xr-x 148 root root 12288 Apr  9 16:48 /etc
$ ls -al /etc | head -n5
total 1268
drwxr-xr-x 148 root root   12288 Apr  9 16:48 .
drwxr-xr-x  21 root root    4096 Mar 17 02:59 ..
drwxr-xr-x   3 root root    4096 Aug  1  2020 acpi
-rw-r--r--   1 root root    3028 Aug  1  2020 adduser.conf
```

Reply:
I see, makes sense. The name of that parameter is a bit awkward, but I guess consistency wins.

Reply:
Good call, renamed to `itself` for clarity.